File: spec/cpp_stl/test_bits_byte_aligned.cpp
#include <boost/test/unit_test.hpp>

#include <bits_byte_aligned.h>

#include <iostream>
#include <fstream>
#include <vector>

BOOST_AUTO_TEST_CASE(test_bits_byte_aligned) {
    std::ifstream ifs("src/fixed_struct.bin", std::ifstream::binary);
    kaitai::kstream ks(&ifs);
    bits_byte_aligned_t* r = new bits_byte_aligned_t(&ks);

    // 50 (6 + 2) = 010100|00
    BOOST_CHECK_EQUAL(r->one(), 0b010100);
    // 41
    BOOST_CHECK_EQUAL(r->byte_1(), 0x41);
    // 43 (3 + 1 + 4) = 010|0|0011
    BOOST_CHECK_EQUAL(r->two(), 0b010);
    BOOST_CHECK_EQUAL(r->three(), false);
    // 4B
    BOOST_CHECK_EQUAL(r->byte_2(), 0x4b);
    // 2D 31 (14 + 2) = 00101101 001100|01
    BOOST_CHECK_EQUAL(r->four(), 0b00101101001100);
    // FF
    BOOST_CHECK_EQUAL(r->byte_3(), "\xff");
    // FF
    BOOST_CHECK_EQUAL(r->full_byte(), 0xff);
    // 50
    BOOST_CHECK_EQUAL(r->byte_4(), 0x50);

    delete r;
}

File: spec/java/src/io/kaitai/struct/spec/TestBitsByteAligned.java
package io.kaitai.struct.spec;

import io.kaitai.struct.testformats.BitsByteAligned;
import org.testng.annotations.Test;

import static org.testng.Assert.assertEquals;

public class TestBitsByteAligned extends CommonSpec {
    @Test
    public void testBitsByteAligned() throws Exception {
        BitsByteAligned r = BitsByteAligned.fromFile(SRC_DIR + "fixed_struct.bin");

        // 50 (6 + 2) = 010100|00
        assertEquals(r.one(), 0b010100);
        // 41
        assertEquals(r.byte1(), 0x41);
        // 43 (3 + 1 + 4) = 010|0|0011
        assertEquals(r.two(), 0b010);
        assertEquals(r.three(), false);
        // 4B
        assertEquals(r.byte2(), 0x4b);
        // 2D 31 (14 + 2) = 00101101 001100|01
        assertEquals(r.four(), 0b00101101_001100);
        // FF
        assertEquals(r.byte3(), new byte[] { (byte) 0xff });
        // FF
        assertEquals(r.fullByte(), 0xff);
        // 50
        assertEquals(r.byte4(), 0x50);
    }
}
